rag_agent.py

```diff
@@ -71,74 +66,72 @@ def setup_vectorstore(self, file_paths):
-            retriever: The document retriever.
-            threshold (float): The confidence threshold to decide
-                               if a query is related to the document.
-    
+            question (str): The user query.
+        
         Returns:
-            result (dict): Retrieved document information and similarity score.
+            bool: True if the query is coding-related, False otherwise.
         """
-        self.retriever = self.setup_vectorstore(document_paths)
-        results = self.retriever.invoke(query)
-        if len(results) > 0:
-            # Check the confidence score of the first result
-            # (assuming retriever returns sorted results)
-            top_result = results[0]
-            score = top_result.metadata.get("score", 0.0)
-    
-            if score >= threshold:
-                return top_result, score
-        return None, 0.0
-    
+        coding_keywords = [
+            'code', 'Python', 'bug', 'error', 'exception', 'syntax', 'Pygame', 'debug', 
+            'program', 'programming', 'variable', 'loop', 'if', 'function', 'print', 
+            'input', 'string', 'number', 'list', 'tuple', 'dictionary', 'module', 
+            'import', 'turtle', 'indentation', 'syntax', 'Pygame', 'math', 'shape', 
+            'for', 'while', 'repeat', 'class', 'object', 'method', 'attribute', 
+            'operator', 'condition', 'true', 'false', 'boolean', 'integer', 'float', 
+            'type', 'index', 'range', 'len', 'append', 'pop', 'remove', 'sort', 
+            'reverse', 'random', 'sleep', 'time', 'pygame', 'GTK', 'logic', 'numbers', 'correct', 'answer'
+        ]
+        return any(keyword.lower() in question.lower() for keyword in coding_keywords)
+
+
     def run(self):
-        if not self.retriever:
-            return
-
-        # This is deprecated according to langchain docs
-        # please migrate it to the new API
-        # https://python.langchain.com/v0.2/docs/versions/migrating_chains/retrieval_qa/
-        rag_chain = RetrievalQA.from_chain_type(
-            llm=self.get_model(),
-            chain_type="stuff",
-            retriever=self.retriever)
-    
+        """Run the main logic of the RAG agent."""
+        
+        # Format documents for context
+        def format_docs(docs):
+            return "\n\n".join(doc.page_content for doc in docs)
+
+        # Create the LCEL chain
+        qa_chain = (
+            {
+                "context": self.retriever | format_docs, 
+                "question": RunnablePassthrough()  
+            }
+            | self.prompt  
+            | self.model   
+            | StrOutputParser()  
+        )
+
         while True:
-            # We'll need to change this as time goes on as we won't
-            # get user input this way.
-            question = input("Enter your question (or type 'exit' to quit): ").strip()
+            question = input("Hey there, let me know your coding doubt").strip()
```

Comment:
Doesn't seem like a great prompt.

Reply:
Actually the prompt would be empty as the child would be adding his query in the text window and then we'll pass it to our model through this, I had added it just for reference.
I can keep it empty or write something like
```What Python magic are we working on today?```
```Let’s code something cool! Ask me!```
Let me know what are your opinions on this

Reply:
If it's meant to be empty, then leave it empty.